### PR TITLE
[MLIR] Add index bitwidth to the DataLayout

### PR DESCRIPTION
**Pull request by gysit**
When importing from LLVM IR the data layout of all pointer types contains an index bitwidth that should be used for index computations. This revision adds a getter to the DataLayout that provides access to the already stored bitwidth. The function returns an optional since only pointer-like types have an index bitwidth. Querying the bitwidth of a non-pointer type returns std::nullopt.

The new function works for the built-in Index type and, using a type interface, for the LLVMPointerType.

### CODE REVIEW

**Review by Dinistro — COMMENTED** [review]
This is looking fine for me, but I would be more comfortable if @ftynse could take a look as well.

File: mlir/include/mlir/Interfaces/DataLayoutInterfaces.h [hunk 1]
```diff
@@ -57,6 +57,13 @@ uint64_t
 getDefaultPreferredAlignment(Type type, const DataLayout &dataLayout,
                              ArrayRef<DataLayoutEntryInterface> params);
 
+/// Default handler for the index bitwidth request. Computes the result for
+/// the built-in index type and dispatches to the DataLayoutTypeInterface for
+/// other types.
+std::optional<uint64_t>
+getDefaultIndexBitwidth(Type type, const DataLayout &dataLayout,
```
[the remark this answers is not in the document]
Comment (ftynse) [reply mlir/include/mlir/Interfaces/DataLayoutInterfaces.h:64]: Wouldn't that just push the burden of checking on the user?

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp [hunk 2]
```diff
@@ -280,10 +280,9 @@ translateDataLayout(DataLayoutSpecInterface attribute,
               uint64_t abi = dataLayout.getTypeABIAlignment(type) * 8u;
               uint64_t preferred =
                   dataLayout.getTypePreferredAlignment(type) * 8u;
-              layoutStream << size << ":" << abi << ":" << preferred;
-              if (std::optional<uint64_t> index = extractPointerSpecValue(
-                      entry.getValue(), PtrDLEntryPos::Index))
-                layoutStream << ":" << *index;
+              uint64_t index = *dataLayout.getTypeIndexBitwidth(type);
```
Comment (ftynse) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:283]: Are we sure this never returns nullopt?

File: mlir/test/Dialect/LLVMIR/layout.mlir [hunk 3]
```diff
@@ -69,24 +74,17 @@ module attributes { dlti.dl_spec = #dlti.dl_spec<
...
     "test.data_layout_query"() : () -> !llvm.ptr<5>
-    // CHECK: alignment = 4
-    // CHECK: alloca_memory_space = 5
-    // CHECK: bitsize = 32
-    // CHECK: global_memory_space = 2
-    // CHECK: preferred = 8
-    // CHECK: program_memory_space = 3
-    // CHECK: size = 4
-    // CHECK: stack_alignment = 128
-    "test.data_layout_query"() : () -> !llvm.ptr<3>
```
Comment (ftynse) [inline mlir/test/Dialect/LLVMIR/layout.mlir:85]: Why is this removed?

File: mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp [hunk 4]
```diff
@@ -345,6 +345,8 @@ TEST(DataLayout, NullSpec) {
   EXPECT_EQ(layout.getTypeABIAlignment(Float16Type::get(&ctx)), 16u);
   EXPECT_EQ(layout.getTypePreferredAlignment(IntegerType::get(&ctx, 42)), 128u);
   EXPECT_EQ(layout.getTypePreferredAlignment(Float16Type::get(&ctx)), 32u);
+  EXPECT_EQ(layout.getTypeIndexBitwidth(Float16Type::get(&ctx)), std::nullopt);
+  EXPECT_EQ(*layout.getTypeIndexBitwidth(IndexType::get(&ctx)), 64u);
```
Comment (ftynse) [inline mlir/unittests/Interfaces/DataLayoutInterfacesTest.cpp:349]: Nit: don't dereference to be consistent with the code below.

**Review by ftynse — APPROVED** [review]

File: mlir/include/mlir/Interfaces/DataLayoutInterfaces.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Dinistro) [reply mlir/include/mlir/Interfaces/DataLayoutInterfaces.h:64]: I would assume that this is only queried for special types in the first place. Still, users might require casting to such an interface even in then.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp (hunk 2 above)
Comment (gysit) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:283]: It goes through the data layout interface of pointer type which has different execution path but as a last result will return the default pointer size. There is also a test where we check that the index bitwidth is exported if the pointer does not specify any (data-layout.mlir).

If we want to be on the safe side I can keep the if but I do not see a code path where this ever executes (until I am proven wrong...).

 Do you understand why there is ptrType and type in this Case of the type switch? I think ptrType could just be renamed to type?

File: mlir/include/mlir/Interfaces/DataLayoutInterfaces.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (gysit) [reply mlir/include/mlir/Interfaces/DataLayoutInterfaces.h:64]: I believe optional makes sense for properties that are not available on all types. We use something similar (Attribute which can be null) for the other optional properties such as stack alignment. That way we do not depend on an interface.

File: mlir/test/Dialect/LLVMIR/layout.mlir (hunk 3 above)
Comment (ftynse) [reply mlir/test/Dialect/LLVMIR/layout.mlir:85]: Ok. May have been a different pointee type when introduced.

**Review by Dinistro — APPROVED** [review]
